kafka_consumer/check.py

```diff
@@ -29,35 +40,161 @@ def __init__(self, name, init_config, agentConfig, instances=None):
         self.kafka_timeout = int(
             init_config.get('kafka_timeout', DEFAULT_KAFKA_TIMEOUT))
 
-    def check(self, instance):
-        consumer_groups = self.read_config(instance, 'consumer_groups',
-                                           cast=self._validate_consumer_groups)
-        zk_connect_str = self.read_config(instance, 'zk_connect_str')
-        kafka_host_ports = self.read_config(instance, 'kafka_connect_str')
+    def _get_highwater_offsets(self, kafka_hosts_ports):
+        """
+        Fetch highwater offsets for each topic/partition from Kafka cluster.
+
+        Do this for all partitions in the cluster because even if it has no
+        consumers, we may want to measure whether producers are successfully
+        producing. No need to limit this for performance because fetching broker
+        offsets from Kafka is a relatively inexpensive operation.
+
+        Sends one OffsetRequest per broker to get offsets for all partitions
+        where that broker is the leader:
+        https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol#AGuideToTheKafkaProtocol-OffsetAPI(AKAListOffset)
+        """
+        # We could switch this to a long-lived conneciton that's cached
+        # on the AgentCheck instance, in which case we don't have to
+        # cleanup the connection. Downside is there's no way to cleanup
+        # connection when agent restarts, so brokers before 0.9 just
+        # accumulate stale connections. In 0.9 Kafka added
+        # connections.max.idle.ms https://issues.apache.org/jira/browse/KAFKA-1282
+        kafka_client = KafkaClient(
+            bootstrap_servers=kafka_hosts_ports,
+            request_timeout_ms=(self.kafka_timeout * 1000))
+        try:
+            highwater_offsets = {}
+            # store partitions that exist but unable to fetch offsets for later
+            # error checking
+            topic_partitions_without_a_leader = []  # contains (topic, partition) tuples
+            for broker in kafka_client.cluster.brokers():
+                if not kafka_client.ready(broker.nodeId):
+                    time.sleep(0.5)
```

Comment:
Yes, this is definitely a hack that I'd rather do differently. 

I also looked at caching the client instance, as well as trying to all the requests asynchronously, but ran into some design tradeoffs. You can see some of my notes on this in the source of #423 where caching `KafkaClient` has many more benefits.

1. The biggest drawback is that there's no way that I could find to trigger connection cleanup when the agent is stopped. Kafka added idle connection reaping in the brokers starting with 0.9. So I'd probably be fine with assuming in #423 that the broker is >= 0.9 as the only broker version that has kafka-based offsets but not idle-connection reaping is 0.8.2. Although it's still a little sketch if someone reports, "I ran #423 but you crashed my production 0.8.2 cluster" 😄 . However, this particular check, which is only fetching zookeeper offsets, is far more likely to be run against old brokers, so I'm hesitant to do that unless there's a clean way to teardown that connection upon graceful shutdown. 

2. Doing it async / caching the connection also adds a fair degree of complexity. It can certainly be done, but I'd rather get a working version live first without it. It's easier to debug if someone hits issues that aren't cropping up in either of our test environments. After 6 months of letting this check sit in production release of the agent, then I'd swap over to async.

3. Re-initializing this every time guarantees we'll have fresh metadata. Otherwise, we may need to trigger a metadata refresh in some scenarios. `KafkaClient` should do this automatically, but when I read the source code, I had just enough uncertainty that it would automatically refresh when I wanted (typically certain error conditions) that I shied away from trusting it. Not opposed to it if you or someone else wants to figure that out or just take a risk that it's doing the right thing. And if it's not doing the right thing, then arguably upstream `kafka-python` should be patched rather than trying to solve it here.

4. Even with all this, ideally we wouldn't have to `sleep` to fake blocking behavior. It'd be simpler to just block until the metadata refresh either finishes or times out. Again, I looked at the source of `kafka-python` for a while, but couldn't figure out how to achieve that. I have a feeling it's possible and I just missed it.

So with all that swirling around, that's why I stayed with the dirty hack. Very open to suggestions.